enterprise/causal-clustering/src/main/java/org/neo4j/causalclustering/core/state/snapshot/PersistentSnapshotDownloader.java

```diff
@@ -0,0 +1,157 @@
+import org.neo4j.causalclustering.core.consensus.LeaderLocator;
+import org.neo4j.causalclustering.core.consensus.NoLeaderFoundException;
+import org.neo4j.causalclustering.core.state.CommandApplicationProcess;
+import org.neo4j.causalclustering.helper.ExponentialBackoffStrategy;
+import org.neo4j.causalclustering.helper.TimeoutStrategy;
+import org.neo4j.logging.Log;
+
+class PersistentSnapshotDownloader implements Runnable
+{
+    static final String OPERATION_NAME = "download of snapshot";
+
+    private final CommandApplicationProcess applicationProcess;
+    private final LeaderLocator leaderLocator;
+    private final CoreStateDownloader downloader;
+    private final Log log;
+    private final TimeoutStrategy.Timeout timeout;
+    private State state;
+
+    PersistentSnapshotDownloader( LeaderLocator leaderLocator,
+            CommandApplicationProcess applicationProcess, CoreStateDownloader downloader, Log log,
+            TimeoutStrategy.Timeout timeout )
+    {
+        this.applicationProcess = applicationProcess;
+        this.leaderLocator = leaderLocator;
+        this.downloader = downloader;
+        this.log = log;
+        this.timeout = timeout;
+        this.state = State.INITIATED;
+    }
+
+    PersistentSnapshotDownloader( LeaderLocator leaderLocator,
+            CommandApplicationProcess applicationProcess, CoreStateDownloader downloader, Log log )
+    {
+        this( leaderLocator, applicationProcess, downloader, log,
+                new ExponentialBackoffStrategy( 1, 30, TimeUnit.SECONDS ).newTimeout() );
+    }
+
+    private enum State
+    {
+        INITIATED,
+        RUNNING,
+        STOPPED,
+        COMPLETED
+    }
+
+    @Override
+    public void run()
+    {
+        if ( !initialStateOk() )
+        {
+            return;
+        }
+        try
+        {
+            state = State.RUNNING;
+            applicationProcess.pauseApplier( OPERATION_NAME );
+            while ( state == State.RUNNING )
+            {
+                try
+                {
+                    downloader.downloadSnapshot( leaderLocator.getLeader() );
+                    break;
+                }
+                catch ( StoreCopyFailedException e )
+                {
+                    log.error( "Failed to download snapshot. Retrying in {} ms.", timeout.getMillis(), e );
+                }
+                catch ( NoLeaderFoundException e )
+                {
+                    log.warn( "No leader found. Retrying in {} ms.", timeout.getMillis() );
+                }
+                Thread.sleep( timeout.getMillis() );
+                timeout.increment();
+            }
+        }
+        catch ( InterruptedException e )
+        {
+            log.error( "Persistent snapshot downloader was interrupted" );
+        }
+        finally
+        {
+            applicationProcess.resumeApplier( OPERATION_NAME );
+            state = State.COMPLETED;
+        }
+    }
+
+    private boolean initialStateOk()
```

Comment:
I don't think you should need this method at all. The state transitions should be obvious and simple. Do you really need INITIATED at all? Can't it just start out in RUNNING-state? Perhaps all you really need is a simple boolean which says isComplete and similarly the method isRunning(), which is only used from tests, can perhaps be eliminated?